### PR TITLE
Add Linq method for getting value and index.

### LINKED ISSUE #95563: [API Proposal]: Linq method for getting value and index
### Background and motivation

There is an existing Linq method for returning an enumerable's value and index, but it was added pre-tuples which makes it clunky as it forces the caller to create a type representation for the combination of the value and index, for example using this:

```C#
IEnumerable<string> lines = GetLines();

foreach (var (line, lineIndex) in lines.Select((l, i) => (l, i)))
{
    // ...
}
```

### API Proposal

```C#
namespace System.Linq;

public static partial class Enumerable
{
    public static IEnumerable<(T Value, int Index)> WithIndex<T>(this IEnumerable<T> source);
}

public static partial class Queryable
{
    public static IQueryable<(T Value, int Index)> WithIndex<T>(this IQueryable<T> source);
}
```

### API Usage

Using `WithIndex`, one can express this idea much more succinctly:

```C#
IEnumerable<string> lines = GetLines();

foreach (var (line, lineIndex) in lines.WithIndex())
{
    // ...
}
```

### Alternative Designs

_No response_

### Risks

None.

### PR DESCRIPTION
**Pull request by RaymondHuy**
Resolve https://github.com/dotnet/runtime/issues/95563

### CODE REVIEW

File: src/libraries/System.Linq/tests/IndexTests.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        [Fact]
+        public void Index_SourceIsNull_ArgumentNullExceptionThrown()
+        {
+            IEnumerable<int> source = null;
+            
+            AssertExtensions.Throws<ArgumentNullException>("source", () => source.Index());
+        }
+
+        [ConditionalFact(typeof(TestEnvironment), nameof(TestEnvironment.IsStressModeEnabled))]
+        public void IndexOverflows()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/IndexTests.cs:20]:
```suggestion
        public void LargeEnumerable_ThrowsOverflowException()
```

File: src/libraries/System.Linq/tests/IndexTests.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var infiniteWhere = new FastInfiniteEnumerator<int>().Index();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/IndexTests.cs:22]: Perhaps constrain down to the smallest possible enumerable that is known to cause this. You also avoid an infinite loop in the hypothetical case that we regress behavior.
```suggestion
            var infiniteWhere = new FastInfiniteEnumerator<int>().Take(int.MaxValue + 2).Index();
```

Comment (eiriktsarpalis) [reply src/libraries/System.Linq/tests/IndexTests.cs:22]: Would it also be possible to add a test showing that this doesn't throw when you take `int.MaxValue + 1` elements?

**Review by eiriktsarpalis — COMMENTED** [review]
There seem to be test failures in CI related to this change.

**Comment by RaymondHuy** [comment]
@eiriktsarpalis  Thanks, I have updated my PR.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

File: src/libraries/System.Linq/src/System/Linq/Index.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        /// <summary>Return index and the associated item.</summary>
```
Comment (stephentoub) [inline src/libraries/System.Linq/src/System/Linq/Index.cs:10]: This isn't an ideal description of what the method does: the method doesn't return an index and an associated item.

File: src/libraries/System.Linq/src/System/Linq/Index.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="source" /> is <see langword="null" />.</exception>
+        public static IEnumerable<(int Index, TSource Item)> Index<TSource>(this IEnumerable<TSource> source)
```
Comment (stephentoub) [inline src/libraries/System.Linq/src/System/Linq/Index.cs:14]: @eiriktsarpalis / @terrajobst, I see ~40 uses in source.dot.net of the Enumerable.Select overload that incorporates an index, 0 of which could switch to use this new method. Why do we believe it's valuable enough to ship?

**Pull request merged**